Environments Documentation (#500)
cloudflare/wrangler: docs/environments.md

## Usage

A common use case that this feature enables is deploying your worker to a staging subdomain before pushing your worker to production. The top-level configuration will be used when running `wrangler publish`. In addition, you can specify multiple environments that you would like to deploy code to. Each environment can inherit properties from the configuration at the top level. `type` will always be inherited from the top-level configuration, you cannot specify different types for different environments. `name` is inherited and modified if left out of the environment configuration, a worker named `my-worker` with an environment `[env.dev]` would become `my-worker-dev`. Fields that can be inherited from the top level configuration are `account_id`, `zone_id`, and `webpack_config`. The rest of the fields must be defined per environment and will not be inherited.

### Examples

#### Base case - routes

This `wrangler.toml` has no environments defined and will publish `my-worker` to `example.com/*`

```toml
type = "webpack"
name = "my-worker"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "example.com/*"

> workers_dot_dev = false // this field specifies that the worker should not be deployed to workers.dev

Reviewer: ```suggestion
workers_dot_dev = false # this field specifies that the worker should not be deployed to workers.dev
```

```

#### Base case - workers.dev

This `wrangler.toml` has no environments defined and will publish `my-worker` to `my-worker.subdomain.workers.dev`

```toml
type = "webpack"
name = "my-worker"
account_id = "12345678901234567890"
workers_dot_dev = true # this field specifies that the worker should be deployed to workers.dev
```

#### Adding Environments

This `wrangler.toml` adds two environments to the base case.

```toml
type = "webpack"
name = "my-worker-dev"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "dev.example.com/*"

[env.production]
name = "my-worker"

> route = "dev.example.com/*"

Author: good catch

```toml
name = "my-worker"
type = "webpack"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "example.com/*"

[env.staging]
workersdotdev = true
```

With this configuration, Wrangler will behave in the following manner:

`wrangler publish` will publish your project to `example.com/*`
`wrangler publish --environment staging` will publish your project to `my-worker-staging.yoursubdomain.workers.dev`

## Invalid configurations

##### Multiple types

You cannot specify a type for each environment, type must be specified at the top level of `wrangler.toml`. 

```toml
name = "my-worker"
type = "webpack"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "example.com/*"

[env.staging]
type = "rust"
workersdotdev = true
```

##### Same name for multiple environments

You cannot specify multiple environments with the same name. If this were allowed, publishing each environment would overwrite your previously deployed worker, and the behavior would not be clear.

```toml
name = "my-worker"
type = "webpack"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "example.com/*"

[env.staging]

> name = "my-worker"

Reviewer 2: Show the error this type of TOML would produce

workersdotdev = true
```

##### Ambiguous top level configuration

```toml
name = "my-worker"
type = "webpack"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "example.com/*
```

You will be warned if `workers_dot_dev` is left out of the top level configuration because if it is not specified, it is unclear what the behavior of `wrangler publish` should be. See [the section on backwards compatibility](#Backwards-compatibility) for more information.

##### Defining workers_dot_dev and route

```toml
name = "my-worker"
type = "webpack"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "example.com/*

[env.staging]
workers_dot_dev = true
route = "staging.example.com/*"
```

> You will be warned if you publish to an environment where `route` is defined and `workers_dot_dev`. Wrangler will publish to `workers.dev`. The reason for the warning is because it is unclear what the intended behavior of the environment should be. Wrangler will assume you mean to deploy to `workers.dev`.

Reviewer 2: Show the warning here too :D

# Environments

Environments is a feature that allows you to deploy the same project to multiple places under multiple names. 

## Usage

> A common use case that this feature enables is deploying your worker to a staging subdomain before pushing your worker to production. The top-level configuration will be used when running `wrangler publish`. In addition, you can specify multiple environments that you would like to deploy code to. Each environment can inherit properties from the configuration at the top level. `type` will always be inherited from the top-level configuration, you cannot specify different types for different environments. `name` is inherited and modified if left out of the environment configuration, a worker named `my-worker` with an environment `[env.dev]` would become `my-worker-dev`. Fields that can be inherited from the top level configuration are `account_id`, `zone_id`, `workers_dot_dev`, and `webpack_config`. `kv_namespaces` and `route` must be defined for each environment and will not be inherited.

Reviewer 3: ```suggestion
Environment's most common use case is deploying your Worker script to a staging subdomain before pushing to production. While`wrangler publish` always runs the top-level configuration, in addition you can specify multiple environments to deploy code to. Each environment can inherit properties from the configuration at the top level with the following caveats:
*  Fields that can be inherited from the top level configuration are `account_id`, `zone_id`, `workers_dot_dev`; `webpack_config`, `kv_namespaces` and `route` must be defined for each environment and will not be inherited.
*  `type` will always be inherited from the top-level configuration; you cannot specify different types for different environments.
* `name` is inherited. If left out of the environment configuration, a Worker project named `my-worker` with an environment `[env.dev]` would become `my-worker-dev`.
```

> ### Concepts

Reviewer 3: I suggest a tiny restructure:
\# Environments
.. include overview of usage here
~~## Usage~~
~~## Concepts~~ Top Level Config
... include caveats on inheritance here

**⚠️ Please note that this behavior will not be supported past v1.5.0!! ⚠️**

A `wrangler.toml` before this release looks like this:

```toml
name = "my-worker"
type = "webpack"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
private = false
route = "example.com/*
```

With this configuration, Wrangler will behave in the following manner:

```console
$ wrangler publish
⚠️  The `private` field is deprecated; please use `workers_dot_dev` to toggle between publishing to your workers.dev subdomain and your own domain.
⚠️  Please specify the workers_dot_dev boolean in the top level of your wrangler.toml.
⚠️  This command will fail in v1.5.0. Please see https://github.com/cloudflare/wrangler/blob/master/docs/environments.md for more information.
✨  Built successfully, built project size is 517 bytes.
✨  Successfully published your script.

> ✨  Success! Your worker was successfully published. You can view it at https://my-worker.subdomain.workers.dev

Reviewer 2: ```suggestion
✨  Success! Your worker was successfully published. You can view it at https://my-worker.<your-subdomain>.workers.dev
```

# Environments

> Environments is a feature that allows you to deploy the same project to multiple places.

Reviewer 4: "under multiple names"

> ## Backwards compatibilty

Reviewer 3: Put Backwards compatibility at the end?

Legacy `wrangler.toml` files will still work as expected during the initial rollout of this feature, however you may notice a deprecation warning. One of the goals of environments is to make it more obvious when you are deploying to a traditional worker with routes, and when you are deploying to a subdomain on workers.dev.

A `wrangler.toml` before this release looks like this:

```toml
name = "my-worker"
type = "webpack"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
private = false
route = "example.com/*
```

With this configuration, Wrangler will behave in the following manner:

`wrangler publish` will publish your worker to your subdomain on workers.dev
`wrangler publish --release` will publish your worker to your route at `example.com/*`. 

It is important to note that both of these commands will issue a deprecation warning. To remove these warnings, you can configure Wrangler with environments to separate deploys to workers.dev from deploys to workers routes.

> ## Intended Use

Reviewer 3: We shouldn't have to say Intended Use.

A common use case that this feature enables is deploying your worker to a staging subdomain before pushing your worker to production. The top-level configuration will be used when running `wrangler publish`. In addition, you can specify multiple environments that you would like to deploy code to.

### Example

```toml
type = "webpack"
name = "my-worker"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "example.com/*"

[env.staging]
name = "my-worker-staging"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "staging.example.com/*"

[env.dev]
name = "my-worker-dev"
account_id = "12345678901234567890"
zone_id = "09876543210987654321"
route = "dev.example.com/*"

> ```

Reviewer: nit: if it was me (and many other very cautious devs out there) i'd likely have prod as a non-default env and dev as my default, making me less likely to fat-finger a production deploy. I'd also recommend this approach to others, even though in reality you could configure this any darn way you please.

also i suppose we should make it clear that you could configure this any darn way you please.

> In order to use environments with this configuration, you can pass the name of the environment via the `--environment` flag. 

Reviewer 4: i think the flag should probably be `--env`

With this configuration, Wrangler will behave in the following manner:

`wrangler publish` will publish your worker to the `example.com/*` route.
`wrangler publish --environment staging` will publish your worker to the `staging.example.com/*` route.

> `wrangler publish --environment dev` will publish your worker to the `dev.example.com/*` route.

Reviewer: i think these would be even clearer as examples with example output; in addition, the output should always indicate where the worker was deployed to.

## Deploying to a workers.dev environment

> In order to deploy your code to workers.dev, you must include `workersdotdev = true` in the desired environment. Your `wrangler.toml` may look like this:

Reviewer: will we enforce that you can have the boolean, the zone/route info, but not both? if so, we should be clear about that here.
